Made some changes to CAT II

johannes-cabal opened the pull request:
Hey, didn't have much time left, but started working from the bottom up. Added the CAT II audit file, included in main.yml, and started hitting some items.

johannes-cabal committed:
Merge pull request #1 from jdspringr07/devel

Devel

johannes-cabal committed:
Started working on CAT II

Added entry to include CAT II in main.yml and added some audits/patches.

johannes-cabal committed:
Tweaked formatting on my changes.

samdoran commented on tasks/audit-cat2.yml:
```diff
@@ -0,0 +1,25 @@
+- name: "MEDIUM | RHEL-07-040640 | AUDIT | The SSH public host key files must have mode 0644 or less permissive."
+  command: find / -name '*ssh_host*key'
```
Please use the find module rather than command.

johannes-cabal replied:
Fixed. Will commit when finished with other changes.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1439,63 +1439,104 @@
 - name: "MEDIUM | RHEL-07-040640 | PATCH | The SSH public host key files must have mode 0644 or less permissive."
-  command: "true"
+  file:
+    dest: "{{ item }}"
```
Please indent four spaces.

johannes-cabal replied:
Fixed. Will commit when finished with other changes.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1439,63 +1439,104 @@
 - name: "MEDIUM | RHEL-07-040650 | PATCH | The SSH private host key files must have mode 0600 or less permissive."
-  command: "true"
+  file:
+    dest: "{{ item }}"
```
Please indent four spaces.

johannes-cabal replied:
Fixed. Will commit when finished with other changes.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1439,63 +1439,104 @@
       - RHEL-07-040700
+      - ssh
 
 - name: "MEDIUM | RHEL-07-040730 | PATCH | The system must not be performing packet forwarding unless the system is a router."
```
Probably should add a variable, e.g.,`rhel7stig_system_is_router`, with a default of `no` and add a conditional to this task. This will allow it to be skipped if the system is a router.

```
when: not rhel7stig_system_is_router
```

johannes-cabal replied:
Fixed, will commit when other changes complete.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1511,12 +1552,15 @@
       - RHEL-07-040740
 
 - name: "MEDIUM | RHEL-07-040810 | PATCH | The system must use a local firewall."
```
Need an additional task that starts and enables the firewall to meet the full requirement.

johannes-cabal replied:
Fixed, will commit when other changes are complete.

samdoran replied:
Actually, this is a duplicate of RHEL-07-040290. Slightly different wording on the requirement but the fix action for both is identical. We can remove these tasks and just leave a comment in line with the finding ID stating it's a duplicate.

Hooray for working from the draft STIG.

johannes-cabal replied:
Yeah, was actually curious if we should leave it now and remove it when the final rev is complete? Not sure which one they will remove. Figured once the final release is out, do a final check for the actual findings, add those that are missing, and remove those that are taken out. 

Thoughts?

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1535,10 +1579,12 @@
+  sysctl:
+      name: net.ipv6.conf.all.accept_source_route
+      present: yes
+      value: 0
```
For `sysctl` tasks, we generally add `ignoreerrors: yes` task parameter to prevent failure due to unknown keys.

johannes-cabal replied:
Fixed on each sysctl task. Getting ready to commit changes now.

johannes-cabal commented:
Yep, will be adding it all in. Mainly wanted to get a commit in to make sure I was following the right styling for stuff.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1439,63 +1439,104 @@
+  sysctl:
+      name: net.ipv4.ip_forward
+      present: yes
+      value: 0
```
Add `ignoreerrors: yes` as well as a conditional to specify whether the system is a router.

johannes-cabal committed:
Made corrective changes to previous commit.

johannes-cabal committed:
Spacing...

johannes-cabal committed:
Made some updates to audit,main,patch, and defaults

- Added & Updated Cat II Audit
- Added CAT II Patches
- Added a default item
- Removed rhel7stig_cat1_patch from main.yml (fails on condition), will
correct later

johannes-cabal committed:
Updated a label.

Fixed a label typo.

johannes-cabal committed:
Corrected finding label

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1440,9 +1440,9 @@
-    dest: "{{ item }}"
-    mode: 0644
-    state: file
+     dest: "{{ item }}"
```
Use `item.path`.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1440,9 +1440,9 @@
+     dest: "{{ item }}"
+     mode: 0644
+     state: file
   with_items: "{{ rhel_07_040640_audit.stdout_lines }}"
```
Use `rhel_07_040640_audit.files`.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1452,10 +1452,10 @@
-    mode: 0600
-    state: file
-  with_items: "{{ rhel_07_040650_audit.stdout_lines }}"
+     dest: "{{ item }}"
```
Use `item.path`.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1452,10 +1452,10 @@
+     dest: "{{ item }}"
+     mode: 0600
+     state: file
+  with_items: "{{ rhel_07_040650_audit.files | map(attribute='path') | list }}"
```
Use `rhel_07_04650.files` without the filters.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1524,6 +1524,7 @@
       regexp: (?i)^#?compression
       line: Compression no
       validate: sshd -t -f %s
+  ignore_errors: yes
```
Why ignore errors for this task?

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1537,6 +1538,8 @@
       name: net.ipv4.ip_forward
       present: yes
       value: 0
+  ignore_errors: yes
```
Use the module parameter `ignoreererors` rather than that task parameter `ignore_errors`. That tells `sysctl` to ignore unknown/invalid keys buts will still allow the task to fail if something else is wrong.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1583,6 +1598,7 @@
       name: net.ipv6.conf.all.accept_source_route
       present: yes
       value: 0
+  ignore_errors: yes
```
Same here. Use `ignorerrors` in the module not the task. (`syscstl` is special in that it has its own `ignoreerrors` parameter).

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1271,12 +1274,56 @@
+      dest: /var/lib/pki-kra/conf/server.xml
+      regexp: ^#?auths.revocationChecking.bufferSize
+      line: auths.revocationChecking.bufferSize=50
+  when: rhel_07_040230_audit.stat.exists == True
```
`== True` is redundant and can be omitted.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1271,12 +1274,56 @@
+      dest: /var/lib/pki-kra/conf/server.xml
+      regexp: ^#?auths.revocationChecking.enabled
+      line: auths.revocationChecking.enabled=true
+  when: rhel_07_040230_audit.stat.exists == True
```
Sam here.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1271,12 +1274,56 @@
+      dest: /var/lib/pki-kra/conf/server.xml
+      regexp: ^#?enableOCSP
+      line: enableOCSP="true"
+  when: rhel_07_040230_audit.stat.exists == True
```
And here.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1271,12 +1274,56 @@
+  lineinfile:
+      dest: /var/lib/pki-kra/conf/server.xml
+      regexp: ^#?enableOCSP
+      line: enableOCSP="true"
```
Not sure if the quoting here will trip up the YAML parser. May need to wrap the whole line in single quotes.

johannes-cabal replied:
I can add the single quotes, but had no issues with testing in 2.1.1. 

(line exists w/ comment, w/out comment, and doesn't exist at all)

EDIT :Added.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1271,12 +1274,56 @@
+      - RHEL-07-040230
+      - pki
+
+- name: "MEDIUM | RHEL-07-040230 | PATCH | The operating system, if using PKI-based authentication, must implement a local cache of revocation data to certificate validation in case of the inability to access revocation information via the network."
```
Are all these tasks for the same finding? If so, they should be combined into a single task and use a loop to iterate over them since breaking them out doesn't give any additional granularity.

```
with_items:
    - regexp: ^#?enableOCSP
      line: 'enableOCSP="true"'
      dest: /var/lib/pki-kra/conf/server.xml
```

And reference the values with `{{ item.dest }}`, `{{ item.regexp }}`, etc.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1271,12 +1274,56 @@
+      dest: /var/lib/pki-kra/conf/server.xml
+      regexp: ^#?ocspCacheSize
+      line: ocspCacheSize="50"
+  when: rhel_07_040230_audit.stat.exists == True
```
Remove `== True`.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      state: present
+      value: 1
+      reload: yes
+  ignore_errors: yes
```
Use `ignoreerrors` on the module not he task.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      state: present
+      value: 1
+      reload: yes
+  ignore_errors: yes
```
Same.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      value: 1
+      sysctl_set: yes
+      reload: yes
+  ignore_errors: yes
```
Same.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      state: present
+      value: 0
+      reload: yes
+  ignore_errors: yes
```
Same.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      state: present
+      value: 0
+      reload: yes
+  ignore_errors: yes
```
Same.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      state: present
+      value: 0
+      reload: yes
+  ignore_errors: yes
```
Same.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1332,210 @@
+      state: present
+      value: 0
+      reload: yes
+  ignore_errors: yes
```
Same.

johannes-cabal replied:
Done.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1423,12 +1562,18 @@
-  command: "true"
+  yum:
+      name:
+          - "@X Windows System"
```
Pretty sure there is some funny business in the yum module when performing a group uninstall. Might need to enumerate packages and shove them in a variable in `vars/main.yml` if that's cleaner.

johannes-cabal replied:
Ok. I'm going to leave it for the time being and make an issue so I can work w/ you to do this correctly. 

As discussed earlier, would probably be best to register the vars from a group request instead of hard coding. Just need to figure out how to approach it. Could use your insight on it. :+1:

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1440,27 +1585,29 @@
       mode: 0644
       state: file
-  with_items: "{{ rhel_07_040640_audit.stdout_lines }}"
+  with_items: "{{ rhel_07_040640_audit.files }}"
```
👍

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1440,27 +1585,29 @@
 - name: "MEDIUM | RHEL-07-040650 | PATCH | The SSH private host key files must have mode 0600 or less permissive."
   file:
-      dest: "{{ item }}"
+      dest: "{{ item.path }}"
```
👍

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1536,15 +1683,17 @@
+      state: present
       value: 0
+      reload: yes
   ignore_errors: yes
```
Use `ignoreerrors` on the module not the task.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1596,11 +1745,13 @@
+      state: present
       value: 0
+      reload: yes
   ignore_errors: yes
```
Use `ignoreerrors` on the module not the task.

samdoran commented on tasks/main.yml:
```diff
@@ -28,21 +28,18 @@
 
 - name: Include CAT I patches
   include: fix-cat1.yml
-  when: rhel7stig_cat1_patch
```
These conditionals so should not be removed. They are what allow remediation based in severity.

johannes-cabal committed:
Made fixes with suggestions, added some vars, and a placeholder for cat3 audits.

johannes-cabal committed:
Made some slight changes.

Created the audit for 040180, 040181, and 04182. Need to get with you to
figure out how we want to handle LDAP key files.

johannes-cabal committed:
Minor text fix (tags)

change ntp to ntpd i ntags

samdoran commented on tasks/audit-cat2.yml:
```diff
@@ -0,0 +1,80 @@
+- name: "MEDIUM | RHEL-07-040230 | AUDIT | The operating system, if using PKI-based authentication, must implement a local cache of revocation data to certificate validation in case of the inability to access revocation information via the network."
+  stat:
+      path: /var/lib/pki-kra/conf/server.xml
+  register: rhel_07_040230_audit
```
Please add `failed_when`, `changed_when`, and `ignore_errors` to all audit tasks.

johannes-cabal replied:
Done.

samdoran commented on tasks/audit-cat2.yml:
```diff
@@ -0,0 +1,80 @@
+      - always
+      - pki
+
+- name: "MEDIUM | RHEL-07-040210 | AUDIT | The operating system must, for networked systems, synchronize clocks with a server that is synchronized to one of the redundant United States Naval Observatory (USNO) time servers, a time server designated for the appropriate DoD network (NIPRNet/SIPRNet), and/or the Global Positioning System (GPS)."
```
Please organize by vuln ID in ascending order.

johannes-cabal replied:
Done. Been working from the bottom up, didn't even think to organize the audits, haha.

samdoran commented on tasks/audit-cat2.yml:
```diff
@@ -0,0 +1,80 @@
+      - RHEL-07-040180
+      - ldap
+
+- name: "MEDIUM | RHEL-07-040180 | AUDIT | The operating system must implement cryptography to protect the integrity of Lightweight Directory Access Protocol (LDAP) authentication communications."
```
Duplicate vuln ID with improper description. Also, the registered variable name doesn't follow convention.

This check may not be necessary if it's in reference to issue #6.

johannes-cabal replied:
Removed. Intended on starting the AV checks, but decided not to.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1224,6 +1231,7 @@
 
 - name: "MEDIUM | RHEL-07-040180 | PATCH | The operating system must implement cryptography to protect the integrity of Lightweight Directory Access Protocol (LDAP) authentication communications."
   command: "true"
+  when: '"yes" in "{{ rhel_07_040180_audit.stdout_lines }}"'
```
No need for double quotes or curly braces around variables in conditionals. Ansible assumes all strings in a conditional statement are variables unless quoted.

`when: "'yes' in rhel_07_040180_audit.stdout_lines"`

johannes-cabal replied:
Tried using that line originally, but would skip due to conditional checking every time. I ended up wrapping it with braces and quotations because that was the only way I could get it to work properly at the time. 

This is on 2.1.0.

samdoran replied:
You need to use `rhel_07_040180_audit.stdout` and it should work as intended.

samdoran commented on tasks/fix-cat2.yml:
```diff
@@ -1285,118 +1329,210 @@
 
-- name: "MEDIUM | RHEL-07-040290 | PATCH | The operating system must enable an application firewall, if available."
-  command: "true"
+- name: "MEDIUM | RHEL-07-040290 | PATCH | The system must use a local firewall."
```
Nothing wrong with this task, but I'm wondering if there are previous findings that manipulate the firewall that may fail if the firewall is not yet installed. Something to check.

johannes-cabal replied:
Not as far as I know. I believe I looked around wondering the same thing. There are also a couple of other tasks that seem strangely dependent on "X" being installed as well. I'm probably gonna start documenting them as I go along.

samdoran commented on defaults/main.yml:
```diff
@@ -2,6 +2,12 @@ rhel7stig_cat1_patch: yes
 
+# These values match patch values by defaults. To override these values,
+# set them in group_vars, host_sars, or with the -e flag via CLI
+rhel7stig_cat1_audit: yes
```
As a default, I would set the audit value to its corresponding patch value. That way they are "linked" by default but they can be controlled separately by the user.

``` yaml
rhel7stig_cat1_audit: "{{ rhel7stig_cat1_patch }}"
rhel7stig_cat2_audit: "{{ rhel7stig_cat2_patch }}"
rhel7stig_cat3_audit: "{{ rhel7stig_cat3_patch }}"
```

johannes-cabal replied:
Done.

samdoran commented on README.md:
```diff
@@ -14,9 +14,13 @@ Role Variables
-| `rhel7stig_cat1_patch` | True | Correct CAT I findings |
-| `rhel7stig_cat2_patch` | False | Correct CAT II findings |
-| `rhel7stig_cat3_patch` | False | Correct CAT III findings |
+| `rhel7stig_cat1_audit` | True  | Audit for CAT I findings      |
```
Please set the default values to `yes` or `no` and wrap them in backticks so they are formatted as code.

```
| `rhel7stig_cat1_audit` | `yes`  | Audit for CAT I findings      |
```

johannes-cabal replied:
Done.

johannes-cabal committed:
Made adjustments based on suggestions.

johannes-cabal committed:
Updated the README

Fixed with backticks.

johannes-cabal committed:
Merge remote-tracking branch 'upstream/devel'

# Conflicts:
#	README.md
#	defaults/main.yml
#	handlers/main.yml

johannes-cabal commented:
Also, I merged our changes together so you shouldn't have to.
